proposals/0000-visible-forall.rst

```diff
@@ -0,0 +1,649 @@
+         |                               |                                  |
+         +-------------------------------+----------------------------------+
+
+     This is a necessity to avoid parsing conflicts. As a consequence, the
+     ``'`` symbol signifies Template Haskell name quotation rather than
+     ``DataKinds`` promotion, and ``*`` is treated as an infix operator
+     regardless of ``-XStarIsType``.
+
+     For the above grammar to work, make ``forall`` a keyword at the term
+     level. This is to be guarded by ``RequiredTypeArguments``, but only for
+     three releases (see point 9 below).
+
+   * In name resolution, we follow the term-level rules that prioritize data
+     constructors over type constructors, and variables over type variables.
+
+     The advantage is that we have less subtle rules this way, so that in ``f
+     T`` and in ``f (T -> T)``, the ``T`` resolves to the same identifier.
+
+     The disadvantage is that in ``\x -> f (forall x. x)``, the ``x``
+     occurrence refers to the lambda-bound variable rather than the
+     ``forall``-bound type variable, even though the ``forall`` binding is
+     closer to the occurrence.
+
+   * Built-in syntax for tuples and lists within types-in-terms is interpreted
+     as in terms. That is, ``[a]`` is a singleton list rather than the type of
+     a list, and ``(a, b)`` is a pair rather than the type of a pair.
+
+   * In type-checking, types-in-terms are rejected outright, unless converted
+     to a type by the "T2T-Mapping" (see point 8 below).
+
+5. Extend the name resolution rules: when looking up a term-level identifier
+   fails, look for a type-level identifier as fallback. As the result, a
+   term-level expression can also contain type variables and type constructors,
+   which will be either rejected or mapped to types (see "T2T-Mapping" below).
+
+6. In a term-level context, a reference to an ambiguous name (such as
+   ``Proxy``) is always resolved in favor of the data constructor, never the
+   type constructor.
+
+7. In a function application ``f e`` for some expressions ``f`` and ``e``, the
+   type of ``f`` affects neither parsing nor renaming of ``e``.
+
+8. In type checking, we alternate between two distinct modes: *checking* and
+   *inference*. This idea, called bidirectional type checking, is presented in
+   more detail in `"A quick look at impredicativity" <https://www.microsoft.com/en-us/research/uploads/prod/2020/01/quick-look-icfp20.pdf>`_.
+
+   * In inference mode, we never infer ``forall x -> t``.
+
+   * In checking mode, in a function application chain ``f e1 e2 e3``, we
+     follow the rules shown in Figure 4 of "A quick look impredicativity",
+     extended as follows::
+
+        rho = t2t(e)
+        G |- sigma[rho/a]; pis  ~>  Theta; phis; rho_r
+        ---------------------------------------------------------  ITVDQ
+        G |- (forall a -> sigma); e,pis  ~>  Theta; phis; rho_r
```

Comment:
> I don't think that you should require readers of this proposal to read another unrelated article to understand the typing rule.

Neither do I, but @simonpj thought it would be nice to provide a more precise specification by building upon the rules in the quick look paper:

https://github.com/ghc-proposals/ghc-proposals/pull/281#issuecomment-698249688

Personally, I don't have a particular preference with regards to the wording, so I'll change it to whatever the committee finds clearer. My current understanding is that Simon prefers to put the rule in the "Proposed Change Specification", not in "Effect and Interactions".

Reply:
Yes!   Proposed Change Specification should, well, specify the change.   Full typing rules would be best of all; but this is much, much better than some arm-waving.

NB: it may not be obvious the relevant typing rules in Fig 4 of the Quick Look paper have *nothing to do* with impredicativity. They are the *baseline* on which the paper builds.  But the approach taken in the baseline is absolutely essential for this proposal to make sense at all.   

Very much part of the spec, nothing to do with interactions.

Reply:
They don't, but they are not particularly standard either. I wasn't suggesting to have no rule here, but a more standard one, possibly with this in the interaction section as: here is how the rules extend to the system presented in the quicklook paper.

It's not a hill I'm willing to die on, however, so let's forget this.